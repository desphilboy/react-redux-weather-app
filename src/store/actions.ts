import { createActions } from "redux-actions";

export const START_API = "START_API";
export const ERROR_API = "ERROR_API";
export const DONE_API = "DONE_API";

export const SET_COUNTRY = "SET_COUNTRY";

export const SET_UNITS = "SET_UNITS";

export const CLEAR_RECORDS = "CLEAR_RECORDS";

export const { startApi, doneApi, errorApi } = createActions({
	START_API: () => ({ data: false, inProgress: true, error: false }), // payload creator
	DONE_API: (data) => ({ data, inProgress: false, error: false }),
	ERROR_API: (error) => ({ data: false, inProgress: false, error }),
});

export const setUnits = (unitsType) => ({
	type: SET_UNITS,
	payload: unitsType,
});

export const setCountry = (country) => ({
	type: SET_COUNTRY,
	payload: country,
});

export const clearRecords = () => ({
	type: CLEAR_RECORDS,
});
