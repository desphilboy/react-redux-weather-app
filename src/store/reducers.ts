import {
   START_API,
   ERROR_API,
   DONE_API,
   SET_COUNTRY,
   SET_UNITS,
   CLEAR_RECORDS,
} from "./actions";

export const INITIAL_STATE = {
   country: "AU",
   units: "metric",

   apiCall: { data: false, error: false, inProgress: false },
   weatherRecords: [],
};

export const apiCall = (state = INITIAL_STATE.apiCall, action) => {
   switch (action.type) {
      case START_API:
      case ERROR_API:
      case DONE_API:
         return { ...state, ...action.payload };
      default:
         return state;
   }
};

export const country = (state = INITIAL_STATE.country, action) => {
   switch (action.type) {
      case SET_COUNTRY:
         return action.payload;
      default:
         return state;
   }
};

export const units = (state = INITIAL_STATE.units, action) => {
   switch (action.type) {
      case SET_UNITS:
         return action.payload;
      default:
         return state;
   }
};

export const weatherRecords = (
   state = INITIAL_STATE.weatherRecords,
   action
) => {
   switch (action.type) {
      case DONE_API:
         return [...state, action.payload.data];
      case CLEAR_RECORDS:
         return [];
      default:
         return state;
   }
};
