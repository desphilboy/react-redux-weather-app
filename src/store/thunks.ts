import { startApi, doneApi, errorApi } from "./actions";

const appId = "ce242e6f1909e3ca32395ead09817fb6";
const baseUrl = "http://api.openweathermap.org/data/2.5/weather";

export const fetchWeather = ({ city = "Melbourne" } = {}) => (
	dispatch,
	getState
) => {
	const { country, units } = getState();

	let fetchurl = `${baseUrl}?q=${city || "Melbourne"},${
		country || "AU"
	}&units=${units || "metric"}&appid=${appId}`;

	dispatch(startApi());
	fetch(fetchurl)
		.then((response) => response.json())
		.then((data) => {
			console.log("data in fetchWeather", data);
			dispatch(doneApi(data));
		})
		.catch((err) => {
			console.log(err);
			dispatch(errorApi(err));
		});
};
