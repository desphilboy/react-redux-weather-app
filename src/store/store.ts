import { createStore, applyMiddleware, combineReducers } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import { apiCall, country, units } from "./reducers";

const reducer = combineReducers({ apiCall, country, units });

const store = createStore(reducer, composeWithDevTools(applyMiddleware(thunk)));

export default store;
