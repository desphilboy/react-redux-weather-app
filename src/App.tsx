import React, { useEffect, Fragment } from "react";

import { AppStyle, BkSv, OptionRow } from "./App.style";
import { connect } from "react-redux";
import { fetchWeather } from "./store/thunks";
import AustralianWeather from "./components/australia-weather/";
import UnitChanger from "./components/change-units";
import CountryChanger from "./components/change-country";
import Titles from "./components/titles/";

const App = ({ ftch, loading, temperature, wind }) => {
  useEffect(() => {
    ftch();
  }, []);

  return (
    <Fragment>
      <BkSv />
      <AppStyle>
        <Titles />
        <OptionRow>
          <UnitChanger />
          <CountryChanger />
        </OptionRow>

        <AustralianWeather />
        {loading && <div> Loading ... </div>}
        {!loading && (
          <Fragment>
            <div>temperature: {temperature}</div>
            <div>
              wind: {wind && wind.speed}, {wind && wind.deg} degree
            </div>
          </Fragment>
        )}
      </AppStyle>
    </Fragment>
  );
}; // App

const mapState = (state) => ({
  loading: state.apiCall.inProgress,
  temperature: state.apiCall.data && state.apiCall.data.temperature,
  wind: state.apiCall.data && state.apiCall.data.wind,
});

const mapDispatch = (dispatch) => ({
  ftch: () => {
    console.log("inside ftch");
    dispatch(fetchWeather());
  },
});

export default connect(mapState, mapDispatch)(App);
