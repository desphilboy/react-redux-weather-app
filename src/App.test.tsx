import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import store from "./store/store.ts";
import App from "./App.tsx";

test("renders learn react link", () => {
	const wrapper = render(
		<Provider store={store}>
			<App loading={false} error={false} />
		</Provider>
	);
	expect(wrapper).toMatchSnapshot();
});
