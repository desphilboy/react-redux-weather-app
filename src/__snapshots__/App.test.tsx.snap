// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gtssRu dcVfTe"
      />
      <div
        class="sc-bdnylx jbPAgA"
      >
        <div
          class="sc-iCoHVE cYQolu"
        >
          <div
            class="sc-fujyUd hpNXfR"
          >
            <img
              alt="weather"
              src="PartlyCloudyIcon.svg"
            />
            <h1>
              Service Victoria Weather
            </h1>
          </div>
        </div>
        <div
          class="sc-dlnjPT eFRjhE"
        >
          <div
            class="sc-jSFkmK bVlpcn"
          >
            <label>
              Unit types:
            </label>
            <select
              name="units"
            >
              <option
                value="metric"
              >
                Metric
              </option>
              <option
                value="imperial"
              >
                Imperial
              </option>
            </select>
          </div>
          <div
            class="sc-gKAblj hSTuob"
          >
            <label>
              Country:
            </label>
            <select
              name="country"
            >
              <option
                value="AU"
              >
                Australia
              </option>
              <option
                value="UK"
              >
                England
              </option>
              <option
                value="NZ"
              >
                New Zealand
              </option>
              <option
                value="IN"
              >
                India
              </option>
              <option
                value="US"
              >
                United States
              </option>
              <option
                value="IR"
              >
                Iran
              </option>
              <option
                value="RU"
              >
                Russia
              </option>
              <option
                value="IT"
              >
                Italy
              </option>
            </select>
          </div>
        </div>
        <div
          class="sc-hKFyIo fxFVqh"
        >
          <div
            class="sc-eCApGN jdRIFg"
          >
            <label
              data-testid="select-city-label"
            >
              Select city from list:
            </label>
            <select
              name="city"
            >
              <option
                value="Sydney"
              >
                Sydney
              </option>
              <option
                value="Melbourne"
              >
                Melbourne
              </option>
              <option
                value="Alice Springs"
              >
                Alice Springs
              </option>
              <option
                value="Adelaide"
              >
                Adelaide
              </option>
              <option
                value="Hobart"
              >
                Hobart
              </option>
              <option
                value="Perth"
              >
                Perth
              </option>
              <option
                value="Canberra"
              >
                Canberra
              </option>
              <option
                value="Brisbane"
              >
                Brisbane
              </option>
            </select>
          </div>
          <div
            class="sc-eCApGN jdRIFg"
          >
            <label>
               or type city's name:
            </label>
            <input
              data-testid="city-name-input"
              name="city"
              type="text"
            />
          </div>
        </div>
        <div>
           Loading ... 
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gtssRu dcVfTe"
    />
    <div
      class="sc-bdnylx jbPAgA"
    >
      <div
        class="sc-iCoHVE cYQolu"
      >
        <div
          class="sc-fujyUd hpNXfR"
        >
          <img
            alt="weather"
            src="PartlyCloudyIcon.svg"
          />
          <h1>
            Service Victoria Weather
          </h1>
        </div>
      </div>
      <div
        class="sc-dlnjPT eFRjhE"
      >
        <div
          class="sc-jSFkmK bVlpcn"
        >
          <label>
            Unit types:
          </label>
          <select
            name="units"
          >
            <option
              value="metric"
            >
              Metric
            </option>
            <option
              value="imperial"
            >
              Imperial
            </option>
          </select>
        </div>
        <div
          class="sc-gKAblj hSTuob"
        >
          <label>
            Country:
          </label>
          <select
            name="country"
          >
            <option
              value="AU"
            >
              Australia
            </option>
            <option
              value="UK"
            >
              England
            </option>
            <option
              value="NZ"
            >
              New Zealand
            </option>
            <option
              value="IN"
            >
              India
            </option>
            <option
              value="US"
            >
              United States
            </option>
            <option
              value="IR"
            >
              Iran
            </option>
            <option
              value="RU"
            >
              Russia
            </option>
            <option
              value="IT"
            >
              Italy
            </option>
          </select>
        </div>
      </div>
      <div
        class="sc-hKFyIo fxFVqh"
      >
        <div
          class="sc-eCApGN jdRIFg"
        >
          <label
            data-testid="select-city-label"
          >
            Select city from list:
          </label>
          <select
            name="city"
          >
            <option
              value="Sydney"
            >
              Sydney
            </option>
            <option
              value="Melbourne"
            >
              Melbourne
            </option>
            <option
              value="Alice Springs"
            >
              Alice Springs
            </option>
            <option
              value="Adelaide"
            >
              Adelaide
            </option>
            <option
              value="Hobart"
            >
              Hobart
            </option>
            <option
              value="Perth"
            >
              Perth
            </option>
            <option
              value="Canberra"
            >
              Canberra
            </option>
            <option
              value="Brisbane"
            >
              Brisbane
            </option>
          </select>
        </div>
        <div
          class="sc-eCApGN jdRIFg"
        >
          <label>
             or type city's name:
          </label>
          <input
            data-testid="city-name-input"
            name="city"
            type="text"
          />
        </div>
      </div>
      <div>
         Loading ... 
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
